bigquery_etl/schema/__init__.py

```diff
@@ -273,6 +279,64 @@ def _traverse(
                             f"Field {prefix}.{field_path} is missing in schema"
                         )
 
+    def _assert_fields_exactly_unionable(
+        self,
+        fields: Sequence[dict[str, Any]],
+        other_fields: Sequence[dict[str, Any]],
+        parent_field_name: Optional[str] = None,
+    ) -> None:
+        """Raise an error if the two lists of fields don't match in terms of field positions, names, and types."""
+        parent_field_note = f" in {parent_field_name}" if parent_field_name else ""
+
+        if len(fields) != len(other_fields):
+            raise SchemaAssertError(
+                f"Different number of fields{parent_field_note} ({len(fields)} vs {len(other_fields)})."
+            )
+
+        for field_number, (field, other_field) in enumerate(
+            zip(fields, other_fields), start=1
+        ):
+            if field["name"] != other_field["name"]:
+                raise SchemaAssertError(
+                    f"Field #{field_number}{parent_field_note} has different names ({field['name']} vs {other_field['name']})."
+                )
```

Comment:
Could this be changed to something like
```suggestion
                errors.append(
                    f"Field #{field_number}{parent_field_note} has different names ({field['name']} vs {other_field['name']})."
                )
                continue
```
with an exception being raised at the end so the error contains all the conflicting columns.  That would result in dupes like `a != b` `b != a`, but it would be convenient to see everything in one go.

e.g. In https://github.com/mozilla/bigquery-etl/pull/6878 there were two of these errors in the struct